### PR TITLE
Register + Login

### PR DESCRIPTION
**Pull request by themonster2015**
This PR adds the following changes to `dev` 👇 :

- Create the Register component.
- Create the Login component.
- Create user state and action reducers in the redux store.
- Create UserDetails component to test what user can see after logging in/signing up.
- Update routes in App component
- Update content in Home component
- Add jwt-decode to decode token.
- Add redux-persist to package.json to persist state on page reload.

### CODE REVIEW

File: src/components/App.js [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
+import Login from './Login';
 import Destinations from './Destinations';
+import UserDetails from './UserDetails';
 
 const App = () => (
   <Router>
-    <Destinations />
+    <Routes>
+      <Route exact path="/" element={<Home />} />
+      <Route path="/register" element={<Register />} />
```
Comment (AleksandarIvezic) [inline src/components/App.js]: [Required]
I believe that we don't need to render this home page. Our home, or root page should display all destinations and it should enable to all users to see destinations. Once the user tries to make a reservation we should redirect him to the login page if he is not logged in already. Link for login we can add to the Navigation, and if a user is not registered we can provide another link to the registration form. Once the user gets logged in Sign in the link will be switched by Sign out.

File: src/components/App.js [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
+      <Route path="/login" element={<Login />} />
+      <Route path="/destinations" element={<Destinations />} />
+      <Route path="/user" element={<UserDetails />} />
+    </Routes>
```
Comment (AleksandarIvezic) [inline src/components/App.js]: [Requred]
I also think that we don't need to render this UserDetails component. Instead, we will create MyReservations, so users can check all reservations he created.

File: src/redux/Auth/auth.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+const authReducer = (state = initialState, action) => {
+  const { type, payload } = action;
+  switch (type) {
+    case REGISTER_USER:
+      return {
+        ...state, logged_in: true, loading: false, token: payload.token, details: payload.details,
+      };
+    case GET_ERRORS:
```
Comment (AleksandarIvezic) [inline src/redux/Auth/auth.js]: [Optional]

One thing that we will provide from the back-end is the admin property of the user, so you can add it like payload.admin into your user state.

File: src/redux/Auth/auth.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      };
+    case SIGN_IN_USER:
+      return {
+        ...state,
+        logged_in: true,
+        loading: false,
+        token: payload.token,
+        details: payload.details,
+        errors: null,
+      };
```
Comment (AleksandarIvezic) [inline src/redux/Auth/auth.js]: [Optional]
Same here. You can add admin: payload.admin.

**Review by AleksandarIvezic — CHANGES_REQUESTED** [review]
Hi @themonster2015 

STATUS: Changes required 🛑 🛑 🛑 

Great job so far. However, I believe that you should do some changes.  Please check the comments below. If you have questions, or you want to discuss about something please write in the comments or reach me on slack.

Cheers!!! 🥂

File: src/components/App.js (hunk 1 above)
Comment (themonster2015) [reply src/components/App.js]: thanks @ShinobiWarior. I've fixed it and keep the homepage to destination page, but for the purpose of testing the user logout flow I will redirect user to the /home link after he logs in or registers and there he can see a link to log out. I will remove the Home component and move the log out link to the nav when it's merged to dev.

File: src/components/App.js (hunk 2 above)
Comment (themonster2015) [reply src/components/App.js]: I've deleted it

File: src/redux/Auth/auth.js (hunk 4 above)
Comment (themonster2015) [reply src/redux/Auth/auth.js]: I've added it

File: src/redux/Auth/auth.js (hunk 3 above)
Comment (themonster2015) [reply src/redux/Auth/auth.js]: I've added it

**Comment by themonster2015** [comment]
For now you can go to /login to see the login page and /register to register

**Review by AleksandarIvezic — APPROVED** [review]
Hi @themonster2015 

### Status: APPROVED ✔️ ✔️ ✔️ 

Great job Yen. Everything looks fine now. You can merge the branch now. 

Cheers!!! 🥂

**Pull request merged**